File: backend/internal/infrastructure/repository/firestore/wod_repository.go
package firestore

import (
	"context"
	"time"

	"melowod/backend/internal/domain/model"

	"cloud.google.com/go/firestore"
)

type WodRepository struct {
	client *firestore.Client
}

func NewWodRepository(client *firestore.Client) *WodRepository {
	return &WodRepository{
		client: client,
	}
}

func (r *WodRepository) Create(wod *model.Wod) error {
	ctx := context.Background()
	
	// Si no hay ID, generamos uno
	if wod.ID == "" {
		ref := r.client.Collection("wods").NewDoc()
		wod.ID = ref.ID
	}

	wod.CreatedAt = time.Now()
	wod.UpdatedAt = time.Now()

	_, err := r.client.Collection("wods").Doc(wod.ID).Set(ctx, wod)
	return err
}

func (r *WodRepository) GetByID(id string) (*model.Wod, error) {
	ctx := context.Background()
	
	doc, err := r.client.Collection("wods").Doc(id).Get(ctx)
	if err != nil {
		return nil, err
	}

	var w model.Wod
	err = doc.DataTo(&w)
	if err != nil {
		return nil, err
	}

	return &w, nil
}

func (r *WodRepository) List() ([]*model.Wod, error) {
	ctx := context.Background()
	
	iter := r.client.Collection("wods").Documents(ctx)
	var wods []*model.Wod

	for {
		doc, err := iter.Next()
		if err != nil {
			break
		}

		var w model.Wod
		err = doc.DataTo(&w)
		if err != nil {
			return nil, err
		}

		wods = append(wods, &w)
	}

	return wods, nil
}

func (r *WodRepository) Update(wod *model.Wod) error {
	ctx := context.Background()
	
	wod.UpdatedAt = time.Now()
	
	_, err := r.client.Collection("wods").Doc(wod.ID).Set(ctx, wod)
	return err
}

func (r *WodRepository) Delete(id string) error {
	ctx := context.Background()
	
	_, err := r.client.Collection("wods").Doc(id).Delete(ctx)
	return err
}

File: functions/src/infrastructure/repositories/WodRepository.ts
import { db } from '../firebase/admin';
import { IWodRepository, Wod, WodFilters } from '../../domain/entities/Wod';
import { CollectionReference, Query, DocumentData } from '@google-cloud/firestore';

export class WodRepository implements IWodRepository {
  private static instance: WodRepository;
  private wodsCollection: CollectionReference;

  private constructor() {
    this.wodsCollection = db.collection('wods');
  }

  public static getInstance(): WodRepository {
    if (!WodRepository.instance) {
      WodRepository.instance = new WodRepository();
    }
    return WodRepository.instance;
  }

  async create(wod: Omit<Wod, 'id' | 'createdAt' | 'updatedAt'>): Promise<string> {
    const now = new Date();
    const wodRef = this.wodsCollection.doc();
    
    await wodRef.set({
      ...wod,
      id: wodRef.id,
      createdAt: now,
      updatedAt: now
    });

    return wodRef.id;
  }

  async getById(id: string): Promise<Wod | null> {
    const wodDoc = await this.wodsCollection.doc(id).get();
    return wodDoc.exists ? (wodDoc.data() as Wod) : null;
  }

  async update(id: string, wod: Partial<Wod>): Promise<void> {
    const now = new Date();
    await this.wodsCollection.doc(id).update({
      ...wod,
      updatedAt: now
    });
  }

  async delete(id: string): Promise<void> {
    await this.wodsCollection.doc(id).delete();
  }

  async list(filters?: WodFilters): Promise<Wod[]> {
    let query: Query<DocumentData> = this.wodsCollection;

    if (filters) {
      if (filters.type) {
        query = query.where('type', '==', filters.type);
      }
      if (filters.difficulty) {
        query = query.where('difficulty', '==', filters.difficulty);
      }
      if (filters.fromDate) {
        query = query.where('createdAt', '>=', filters.fromDate);
      }
      if (filters.toDate) {
        query = query.where('createdAt', '<=', filters.toDate);
      }
      if (filters.limit) {
        query = query.limit(filters.limit);
      }
    }

    const snapshot = await query.get();
    return snapshot.docs.map(doc => ({
      ...doc.data(),
      id: doc.id
    }) as Wod);
  }
}

export const wodRepository = WodRepository.getInstance();
